### PR TITLE
C++: Basic input iterator models

### PR DESCRIPTION
**Pull request by rdmarsh2**
This adds taint and data flow edges through overloaded operators for input iterators, as well as extending the `std::string` model to include some iterator-related functions. It does not account for writes via output iterators. I'm going to add some additional tests and documentation, and rebase once #4090 is merged.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:202]: I think this is assumed for member function calls?  (I actually need to block this flow for `clear` methods)

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:39]: I think `LegacyIterator` could do with a better name, as it appears to include all the things we can deduce to be iterators.

**Review by geoffw0 — COMMENTED** [review]
Quite a few comments (I appreciate that this is a draft PR).  I get the impression the resulting taint flows we get from this are going to be quite valuable.

File: cpp/ql/src/semmle/code/cpp/models/implementations/StdString.qll
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/models/implementations/StdString.qll:88]: I didn't realize the `end` iterator was random access. Adding that now.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:39]: C++20 has a concept for iterators, which is supposed to replace the older definitions (`LegacyInputIterator` and the like). Since we don't have support yet, all the things we can deduce to be iterators are `LegacyIterator`s. We can call this `Iterator` for now, and break it out to cover legacy vs concept iterators if we need that distinction in the future.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:15]: There might be an issue with partial specialization - I need to think about it a bit more.

Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:15]: I've used something pretty similar to this pattern myself, and I'm gaining confidence but am still a little nervous of it - if you do find a problem I'd like to hear about it.

File: cpp/ql/test/library-tests/dataflow/taint-tests/stl.h
Comment (jbj) [reply cpp/ql/test/library-tests/dataflow/taint-tests/stl.h:75]: I'll take a look. My suspicion right now is on `EscapesTree.qll`.

Comment (jbj) [reply cpp/ql/test/library-tests/dataflow/taint-tests/stl.h:75]: @rdmarsh2 The extractor team figured out the issue. When modelling C++20 libraries, we should configure our qltest for the latest compilers and language revisions. See https://github.com/github/codeql/pull/4122.

Comment (geoffw0) [reply cpp/ql/test/library-tests/dataflow/taint-tests/stl.h:75]: Other versions of the C++ standard have it exactly like this but without the `constexpr`, so I'm quite happy with or without it for the purposes of testing `basic_string`.

File: cpp/ql/src/semmle/code/cpp/models/implementations/StdString.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/StdString.qll:117]: I don't think this change has been pushed.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:1]: Most of the classes have QLDoc now, but it's still missing from the top of the file itself, `IteratorArrayOperator`, and `StdStringBegin`.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:49]: Happy with `Iterator`.

File: cpp/ql/test/library-tests/dataflow/taint-tests/string.cpp
Comment (geoffw0) [reply cpp/ql/test/library-tests/dataflow/taint-tests/string.cpp]: The new tests for string iterators look good to me, and cover the iterator operations well.

Which of `StdIterator`, `IteratorByTypedefs` and `IteratorTraits` do the tests cover?

Comment (geoffw0) [reply cpp/ql/test/library-tests/dataflow/taint-tests/string.cpp]: (tests have been added)

Great, thanks!

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:8]: :+1:

File: cpp/ql/src/semmle/code/cpp/models/implementations/StdString.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/StdString.qll:134]: :+1:

**Review by geoffw0 — APPROVED** [review]
All LGTM now.

@jbj do you have any remaining concerns?

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:67]: Is the argument always going to be an iterator, or could it also be a (const) reference to an iterator? I can see that https://en.cppreference.com/w/cpp/language/operator_member_access gives `T a` in its example rather than `T &a`, but can't there be cases where copying the iterator is expensive? For example, an iterator over a tree with no parent pointers has to maintain a stack.

**Comment by geoffw0** [comment]
There are merge conflicts (probably because https://github.com/github/codeql/pull/4125 was merged - sorry!)

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:66]: `getExplicitlyConverted()`, perhaps? Or `.getConversion*()`, but I'm unsure whether these have good performance.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:84]: Many of these models could be more concise if they used QL's set literal syntax. For example, these four lines could be one line: `this.hasName(["operator++", "operator--"]) and`. Whether you prefer one or the other is of course a matter of personal taste.

There's a similar opportunity for `input.isParameter([0, 1])` below.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:101]: AFAICT from https://en.cppreference.com/w/cpp/language/operator_member_access, there is no non-member version of this operator.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:198]: AFAICT from https://en.cppreference.com/w/cpp/language/operator_member_access there is no non-member version of this operator.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:141]: Aren't these two alternatives identical? If not, please add an explanatory comment.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:168]: I don't agree that a user-controlled integer `i` transfers taint to `it - i`. Either `i` is out of range, in which case there's UB already at the subtraction, or `i` is in range, in which case `it - i` is tainted iff `it` is tainted.

Same for `operator+`, `operator+=` and `operator-=` in their various member/non-member forms.

For the `operator-` variant that returns an integer, I'd argue that the integer is never tainted -- if there'd been anything wrong with its inputs, like coming from different underlying objects, the subtraction itself would be UB.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:262]: Why no taint from the qualifier object?

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:306]: Now that https://github.com/github/codeql/pull/4125/commits/c083c6235debcf041c2373e89a98b41ec6d60e92 is merged, it's appropriate to add taint or data flow from `isReturnValueDeref` to `isQualifierObject` for all functions that return `this`.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:322]: Like for `+` etc., I don't think there is taint from parameter 0.

**Review by jbj — COMMENTED** [review]
Sorry for being a bit late to the party. I've reviewed up to `Iterator.qll` and have these comments so far.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:67]: I think an iterator that isn't lightweight would be a bit strange, but I'm not against making this more robust.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:84]: This syntax is growing on me quite quickly, since @jbj pressured me to try it out. ;)

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:168]: Adding or subtracting a user-defined integer could be dangerous in itself, but I agree it's probably the job of a query to flag that, rather than the job of taint flow to try and guess the consequences.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:141]: It should be argument 0 or argument 1, but the rewrite is a bit more complicated because we don't want flow from the integer argument.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:262]: Fixed.

**Comment by geoffw0** [comment]
I'm away tomorrow, but I've said everything I intend to say and would be very happy to find this merged when I come back on Monday morning (I have some work to do that will require it).  So @jbj once your concerns have been addressed, please don't wait for my approval!

:+1:

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:66]: autoformat needed

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:113]: For the reference case of `calledWithIteratorArgument`, this should be `isParameterDeref`. Same below. Perhaps that can be achieved by changing the `int` parameter of `calledWithIteratorArgument` to be a `FunctionInput`.

**Review by jbj — COMMENTED** [review]
I've marked my previous review comments as resolved where appropriate and added a few new ones.

**Comment by jbj** [comment]
I've resolved all review comments except two.

**Review by jbj — APPROVED** [review]
LGTM. I'll merge now this to unblock other work, and then I'll start a CPP-Differences job.

**Comment by jbj** [comment]
Started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1382/

**Comment by jbj** [comment]
@rdmarsh2 The CPP-Differences show a 5% slowdown on Wireshark, focused around a handful of queries. Can you find out if it's real or noise? If it's real, could it be due to an increased usage of `getQualifiedName`? Callers of that predicate tend to be slower than they ought to, and I haven't fully investigated what can be done about it.

**Comment by geoffw0** [comment]
I've created https://github.com/github/codeql-c-analysis-team/issues/137 to discuss this and related performance worries.

**Comment by rdmarsh2** [comment]
In Wireshark, the slowdown mostly results from evaluating `Expr::getExplicitlyConverted` without any constraints pushed into it. There's a couple queries with slowdowns in the `select` stage as well, but I haven't found any tuple count explanation for them. It might be related to generating more strings, or just noise.

**Comment by rdmarsh2** [comment]
> There's a couple queries with slowdowns in the select stage as well
I didn't find an explanation for similar slowdowns in those queries on Kamailio, but it looks like the Wireshark slowdowns do have tuple count changes - they seem to be rescanning a few predicates that were flushed to disk, and those rescans are attributed to the select stage. I suspect they were evicted because the transitive closure in `Expr::getExplicitlyConverted` created memory pressure.

**Pull request merged**
